text editor with yjs (#3)
partykit/templates: templates/yjs-editor/README.md

# 🎈 Template: yjs-text-editor

Welcome to the party, pal!

> This is a [Partykit](https://partykit.io) project, which lets you create real-time collaborative applications with minimal coding effort.

Reviewer: ```suggestion
This is a [PartyKit](https://partykit.io) project, which lets you create real-time collaborative applications with minimal coding effort.
```

This is an end-to-end example of:

- a multiplayer chat room

> - with AI-generated replies using either OpenAI or Llama2.

Reviewer: This is from the chat-room example?

It's built with client-side React and a PartyKit server, and it's a good starting point for building your own PartyKit project.

Refer to our docs for more information: https://github.com/partykit/partykit/blob/main/README.md. For more help, reach out to us on [Discord](https://discord.gg/g5uqHQJc3z), [GitHub](https://github.com/partykit/partykit), or [X/Twitter](https://twitter.com/partykit_io).

![image](assets/yjs-editor.png)

## How it works

### What we we looking at?

Let's bring up the UI by installing this example and running the code:

```bash
npm create partykit@latest --template yjs-text-editor
cd chat-room

> npm install

Reviewer: ```suggestion
npm create partykit@latest my-yjs-project --template yjs-text-editor
cd my-yjs-project
```

Now we know we can serve a client app, let's create the editor.

This is a standard React app and we won't spend much time on it here.

Have a look at `app/components/Editor.tsx`. This is the component that shows the text editor.

```typescript
import ReactQuill, { Quill } from "react-quill";
import { QuillBinding } from "y-quill";
import useYProvider from "y-partykit/react";

export default function Editor(/*...*/) {
  // ...
  const provider = useYProvider({
    room,
    options: {},
  });

  // Create an editor-binding which
  // "binds" the quill editor to a Y.Text type.
  useEffect(() => {
    // ...
    const ytext = provider.doc.getText("quill");
    const editor = quill.current.getEditor();
    const binding = new QuillBinding(ytext, editor, provider.awareness);
  }, [/* ... */]);

  return (
      <ReactQuill
        ref={quill}
        /* ... */
      />
  );
}
```

Most of the code has been removed so we can zoom out and see the structure.

- [Quill](https://quilljs.com) is a popular rich text editor framework for the web. The library [ReactQuill](https://github.com/zenoamaro/react-quill) is a React wrapper for it. _We're using this to provide the text editor UI._

> - [Yjs](https://yjs.dev) is a popular shared datastructures framework based on CRDTs and particularly good for text. It allows many clients work to together on the same data, and resolves conflicts.

Reviewer: ```suggestion
- [Yjs](https://yjs.dev) is a popular shared datastructures framework based on CRDTs and particularly good for text. It allows many clients to work together on the same data, and resolves conflicts automatically.
```

    );
  }

  // Always return a Response
  return Response.json({ error: "Method not allowed" }, { status: 405 });
}
```

We'll be using `onRequest` in the next section.

### Tracking occupancy

Let's create that occupancy tracker.

Here's the overall architecture of what we're going to build.

![image](assets/architecture.png)

We're building the `rooms` party in the lower right.

#### 1️⃣ Storing occupancy data

The `rooms` party only uses a single room. To make that clear, we'll make a note of the single room name in the const `SINGLETON_ROOM_ID`. (This const is imported by the client component so it knows what room to connect to.)

PartyKit server instances are stateful, so we can store data in the server instance and it will persist between requests.

Let's just keep it in memory for this demo.

```typescript
export const SINGLETON_ROOM_ID = "index";

export default class OccupancyServer implements Party.Server {
  // Track room occupancy
  rooms: Rooms;

  constructor(public room: Party.Room) {
    this.rooms = {};
  }

  /* everything else */
}
```

Ok that's a place to store occupancy data.

> [!TIP]

> > PartyKit servers are stateful but short-running. If you need to store data for longer, as you would in production, there is first-class support for persisting small values. See: [Persisting state into storage](https://docs.partykit.io/guides/persisting-state-into-storage/)

Reviewer: ```suggestion
> PartyKit servers are stateful but can shut down when there's no activity. If you need to store data for longer, as you would in production, there is first-class support for persisting small values. See: [Persisting state into storage](https://docs.partykit.io/guides/persisting-state-into-storage/)
```

Let's jump back to `party/server.ts` and look at just the relevant lines:

```typescript
export default class EditorServer implements Party.Server {
  // ...

  async onConnect(conn: Party.Connection) {
    await this.updateCount(); // <- OCCUPANCY COUNT CHANGED!
    return onConnect(conn, this.room, this.getOpts());
  }

  async onClose(_: Party.Connection) {
    await this.updateCount(); // <- OCCUPANCY COUNT CHANGED!
  }

  async updateCount() {
    // Count the number of live connections
    const count = [...this.room.getConnections()].length;
    // Send the count to the 'rooms' party using HTTP POST
    await this.room.context.parties.rooms.get(SINGLETON_ROOM_ID).fetch({
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ room: this.room.id, count }),
    });
  }
}
```

What's going on?

- Whenever a client WebSocket connects (special method `onConnect`) or closes (special method `onClose`), the server calls `updateCount`.
- `updateCount` counts the number of live connections to the room. `this.room.getConnections()` is an iterable of all connections to the current room. We just want to the length.

> - It then gets the `rooms` party in the current project, and the singleton room, using the _context_ provides by PartyKit: `this.room.context.parties.rooms.get(SINGLETON_ROOM_ID)`.

Reviewer: ```suggestion
- It then gets the `rooms` party in the current project, and the singleton room, using the _context_ provided by PartyKit: `this.room.context.parties.rooms.get(SINGLETON_ROOM_ID)`.
```
